In [152]:
import pandas as pd
import numpy as np

In [153]:
df = pd.read_csv('preprocesseddf.csv')
df.head()

Reason1  Reason2  Reason3  Reason4  Year  Month  Day_of_the_week  \
0        0        0        0        1  2015      7                1   
1        1        0        0        0  2015      7                1   
2        0        0        0        1  2015      7                2   
3        1        0        0        0  2015      7                3   
4        0        0        0        1  2015      7                3   

   Transportation Expense  Distance to Work  Age  Daily Work Load Average  \
0                     289                36   33                  239.554   
1                     118                13   50                  239.554   
2                     179                51   38                  239.554   
3                     279                 5   39                  239.554   
4                     289                36   33                  239.554   

   Body Mass Index  Education  Children  Pets  Absenteeism Time in Hours  
0               30          0         2     1                          4  
1               31          0         1     0                          0  
2               31          0         0     0                          2  
3               24          0         2     0                          4  
4               30          0         2     1                          2

# Targets

First, we must decide that is this a classification or regression problem, if we refer the the absent hour time feature, it is filled with the amount of hour certain employee has been absent, if we are going to predict that, this mean that this is a regression problem.

Alternatively, we can turn this into a classification, from the amount of time absent, we can decide how many is excessive and how many is normal, many way to decide this, but to make it simple, lets refer to the median.

Lets decide that this is a classification problem so we have to turn the amount of hours into classes of excesive or not.

In [154]:
df.describe()

Reason1     Reason2     Reason3     Reason4         Year      Month  \
count  700.000000  700.000000  700.000000  700.000000   700.000000  700.00000   
mean     0.304286    0.008571    0.090000    0.597143  2016.512857    6.36000   
std      0.460433    0.092250    0.286386    0.490823     0.972353    3.50501   
min      0.000000    0.000000    0.000000    0.000000  2015.000000    1.00000   
25%      0.000000    0.000000    0.000000    0.000000  2016.000000    3.00000   
50%      0.000000    0.000000    0.000000    1.000000  2016.000000    6.00000   
75%      1.000000    0.000000    0.000000    1.000000  2017.000000   10.00000   
max      1.000000    1.000000    1.000000    1.000000  2018.000000   12.00000   

       Day_of_the_week  Transportation Expense  Distance to Work         Age  \
count       700.000000              700.000000        700.000000  700.000000   
mean          2.011429              222.347143         29.892857   36.417143   
std           1.480396               66.312960         14.804446    6.379083   
min           0.000000              118.000000          5.000000   27.000000   
25%           1.000000              179.000000         16.000000   31.000000   
50%           2.000000              225.000000         26.000000   37.000000   
75%           3.000000              260.000000         50.000000   40.000000   
max           6.000000              388.000000         52.000000   58.000000   

       Daily Work Load Average  Body Mass Index   Education    Children  \
count               700.000000       700.000000  700.000000  700.000000   
mean                271.801774        26.737143    0.167143    1.021429   
std                  40.021804         4.254701    0.373370    1.112215   
min                 205.917000        19.000000    0.000000    0.000000   
25%                 241.476000        24.000000    0.000000    0.000000   
50%                 264.249000        25.000000    0.000000    1.000000   
75%                 294.217000        31.000000    0.000000    2.000000   
max                 378.884000        38.000000    1.000000    4.000000   

             Pets  Absenteeism Time in Hours  
count  700.000000                 700.000000  
mean     0.687143                   6.761429  
std      1.166095                  12.670082  
min      0.000000                   0.000000  
25%      0.000000                   2.000000  
50%      0.000000                   3.000000  
75%      1.000000                   8.000000  
max      8.000000                 120.000000

In [155]:
df.iloc[:, -1].median()

3.0

From that, we know that majority of amount of absent time is 3 Hours. lets decide that this is the cut-off point from normal amount and excesive amount.

In [156]:
df['Target'] = np.where(df.iloc[:, -1] <= 3, 0, 1)
df['Target']

0      1
1      0
2      0
3      1
4      0
      ..
695    1
696    0
697    1
698    0
699    0
Name: Target, Length: 700, dtype: int32

In [157]:
df

Reason1  Reason2  Reason3  Reason4  Year  Month  Day_of_the_week  \
0          0        0        0        1  2015      7                1   
1          1        0        0        0  2015      7                1   
2          0        0        0        1  2015      7                2   
3          1        0        0        0  2015      7                3   
4          0        0        0        1  2015      7                3   
..       ...      ...      ...      ...   ...    ...              ...   
695        1        0        0        0  2018      5                2   
696        1        0        0        0  2018      5                2   
697        1        0        0        0  2018      5                3   
698        0        0        0        1  2018      5                3   
699        0        0        0        1  2018      5                3   

     Transportation Expense  Distance to Work  Age  Daily Work Load Average  \
0                       289                36   33                  239.554   
1                       118                13   50                  239.554   
2                       179                51   38                  239.554   
3                       279                 5   39                  239.554   
4                       289                36   33                  239.554   
..                      ...               ...  ...                      ...   
695                     179                22   40                  237.656   
696                     225                26   28                  237.656   
697                     330                16   28                  237.656   
698                     235                16   32                  237.656   
699                     291                31   40                  237.656   

     Body Mass Index  Education  Children  Pets  Absenteeism Time in Hours  \
0                 30          0         2     1                          4   
1                 31          0         1     0                          0   
2                 31          0         0     0                          2   
3                 24          0         2     0                          4   
4                 30          0         2     1                          2   
..               ...        ...       ...   ...                        ...   
695               22          1         2     0                          8   
696               24          0         1     2                          3   
697               25          1         0     0                          8   
698               25          1         0     0                          2   
699               25          0         1     1                          2   

     Target  
0         1  
1         0  
2         0  
3         1  
4         0  
..      ...  
695       1  
696       0  
697       1  
698       0  
699       0  

[700 rows x 17 columns]

Target :
* 0 = No absent or normal amount of absent
* 1 = Excessive

In [158]:
df['Target'].value_counts(normalize=True)

0    0.544286
1    0.455714
Name: Target, dtype: float64

In [159]:
#drop absent time hours and Year
df2 = df.drop(['Absenteeism Time in Hours', 'Year'], axis =1)
df2

Reason1  Reason2  Reason3  Reason4  Month  Day_of_the_week  \
0          0        0        0        1      7                1   
1          1        0        0        0      7                1   
2          0        0        0        1      7                2   
3          1        0        0        0      7                3   
4          0        0        0        1      7                3   
..       ...      ...      ...      ...    ...              ...   
695        1        0        0        0      5                2   
696        1        0        0        0      5                2   
697        1        0        0        0      5                3   
698        0        0        0        1      5                3   
699        0        0        0        1      5                3   

     Transportation Expense  Distance to Work  Age  Daily Work Load Average  \
0                       289                36   33                  239.554   
1                       118                13   50                  239.554   
2                       179                51   38                  239.554   
3                       279                 5   39                  239.554   
4                       289                36   33                  239.554   
..                      ...               ...  ...                      ...   
695                     179                22   40                  237.656   
696                     225                26   28                  237.656   
697                     330                16   28                  237.656   
698                     235                16   32                  237.656   
699                     291                31   40                  237.656   

     Body Mass Index  Education  Children  Pets  Target  
0                 30          0         2     1       1  
1                 31          0         1     0       0  
2                 31          0         0     0       0  
3                 24          0         2     0       1  
4                 30          0         2     1       0  
..               ...        ...       ...   ...     ...  
695               22          1         2     0       1  
696               24          0         1     2       0  
697               25          1         0     0       1  
698               25          1         0     0       0  
699               25          0         1     1       0  

[700 rows x 15 columns]

Separate inputs and target(output)

# Splitting

To train ML model, it is advised to not train it with the whole data, instead make 2 (sometimes 3) sets so we have a set for training the model, then test the model with another set that is not used to train, in other word, the model has not seen this portion of the data and will make a prediction on this data.

In [160]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(df2.drop('Target', axis=1), df2.Target, test_size = 0.25, stratify = df2.Target)

# Standardize

Notice that our numerical data is varied by ranges, before we use this data for our machine learning model training, we have to standardize it so feature with huge number will not affect the model to be biased toward that feature.

In [161]:
from sklearn.preprocessing import StandardScaler

In [162]:
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

We scale the test set using the training set attribute, hence, no fitting.

# Modeling: Logistic Regression

Finally, we come to the actual ML part, because we already decide that we want to predict that the absent amount is Excessive(noted by '1') OR not(noted by '0'). we decide to use Logistic Regression as our estimator

In [163]:
from sklearn.linear_model import LogisticRegression

In [164]:
logreg = LogisticRegression()
logreg.fit(X_train_scaled, y_train)

LogisticRegression()

In [165]:
logreg.score(X_train_scaled, y_train)

0.7371428571428571

Our model accuracy is : 71%
but this is on training set, lets test it on test set

## Intercept and Coeficient

In [166]:
logreg.intercept_

array([-0.16846896])

In [167]:
logreg.coef_

array([[ 0.22469421, -0.04904681,  0.42231866, -0.44267411, -0.03150407,
        -0.11854605,  0.63726104, -0.12449875, -0.38891351,  0.00219199,
         0.3681015 , -0.12204579,  0.24018961, -0.36131155]])

In [168]:
featurename = X_train.columns.values

In [169]:
sumtable = pd.DataFrame(columns = ['Feature Name'], data = featurename)
sumtable['Coef'] = np.transpose(logreg.coef_)
sumtable

Feature Name      Coef
0                   Reason1  0.224694
1                   Reason2 -0.049047
2                   Reason3  0.422319
3                   Reason4 -0.442674
4                     Month -0.031504
5           Day_of_the_week -0.118546
6    Transportation Expense  0.637261
7          Distance to Work -0.124499
8                       Age -0.388914
9   Daily Work Load Average  0.002192
10          Body Mass Index  0.368102
11                Education -0.122046
12                 Children  0.240190
13                     Pets -0.361312

In [170]:
sumtable.index = sumtable.index+1
sumtable.loc[0] = ['Intercept', logreg.intercept_[0]]

In [171]:
sumtable.sort_index(inplace=True)
sumtable

Feature Name      Coef
0                 Intercept -0.168469
1                   Reason1  0.224694
2                   Reason2 -0.049047
3                   Reason3  0.422319
4                   Reason4 -0.442674
5                     Month -0.031504
6           Day_of_the_week -0.118546
7    Transportation Expense  0.637261
8          Distance to Work -0.124499
9                       Age -0.388914
10  Daily Work Load Average  0.002192
11          Body Mass Index  0.368102
12                Education -0.122046
13                 Children  0.240190
14                     Pets -0.361312

# Interpreting Coefficients

In [172]:
sumtable['Odds_ratio'] = np.exp(sumtable.Coef)
sumtable

Feature Name      Coef  Odds_ratio
0                 Intercept -0.168469    0.844957
1                   Reason1  0.224694    1.251940
2                   Reason2 -0.049047    0.952137
3                   Reason3  0.422319    1.525495
4                   Reason4 -0.442674    0.642316
5                     Month -0.031504    0.968987
6           Day_of_the_week -0.118546    0.888211
7    Transportation Expense  0.637261    1.891294
8          Distance to Work -0.124499    0.882939
9                       Age -0.388914    0.677793
10  Daily Work Load Average  0.002192    1.002194
11          Body Mass Index  0.368102    1.444989
12                Education -0.122046    0.885108
13                 Children  0.240190    1.271490
14                     Pets -0.361312    0.696762

In [173]:
sumtable.sort_values('Odds_ratio', ascending=False)

Feature Name      Coef  Odds_ratio
7    Transportation Expense  0.637261    1.891294
3                   Reason3  0.422319    1.525495
11          Body Mass Index  0.368102    1.444989
13                 Children  0.240190    1.271490
1                   Reason1  0.224694    1.251940
10  Daily Work Load Average  0.002192    1.002194
5                     Month -0.031504    0.968987
2                   Reason2 -0.049047    0.952137
6           Day_of_the_week -0.118546    0.888211
12                Education -0.122046    0.885108
8          Distance to Work -0.124499    0.882939
0                 Intercept -0.168469    0.844957
14                     Pets -0.361312    0.696762
9                       Age -0.388914    0.677793
4                   Reason4 -0.442674    0.642316

Interpretation : for each increased unit in features, the odds will be multiplied by the value of odds_ratio.

For example : if we have an odds of (our intercept is 0.83, lets round it up for simplification)5:1, and our employee is absent with reason3, the odds is now 8.12:1, increasing the odds of absent being excessive(from 5:1), on the other hand, if employee is absent with reason4, the odds is now 3.45:1 decreasing the odds of absent being excessive(from 5:1). Same rules applied for other features including numerical but remember that we standardize the numerical feature, so, for example rather than for each unit, for example 'Dialy Work Load Average', its standardize unit of 'Dialy Work Load Average'.

Can the model be improved? What can we do?
well yes, when we standardize the data, we also standardize the dummies which only contains 0 and 1, in other words its not actually a continuous features, we should not do that in the first place.

In [182]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(df2.drop(['Target', 'Reason4'], axis=1), df2.Target, test_size = 0.25, stratify = df2.Target)

In [183]:
from sklearn.compose import make_column_transformer
df2.columns

Index(['Reason1', 'Reason2', 'Reason3', 'Reason4', 'Month', 'Day_of_the_week',
       'Transportation Expense', 'Distance to Work', 'Age',
       'Daily Work Load Average', 'Body Mass Index', 'Education', 'Children',
       'Pets', 'Target'],
      dtype='object')

In [184]:
#Select Features to be transformed (leave all categorical features)
coltotrans = ['Month', 'Day_of_the_week',
       'Transportation Expense', 'Distance to Work', 'Age',
       'Daily Work Load Average', 'Body Mass Index', 'Children',
       'Pets']
ct = make_column_transformer([StandardScaler(), coltotrans], remainder='passthrough')
X_train_scaled = ct.fit_transform(X_train)

In [185]:
X_train_scaled = pd.DataFrame(X_train_scaled, columns = ['Month', 'Day_of_the_week',
       'Transportation Expense', 'Distance to Work', 'Age',
       'Daily Work Load Average', 'Body Mass Index', 'Children',
       'Pets', 'Reason1', 'Reason2', 'Reason3', 'Education'])
X_train_scaled

Month  Day_of_the_week  Transportation Expense  Distance to Work  \
0   -0.075946        -0.682604               -1.575780         -1.154343   
1   -0.360745         0.672281                1.590516         -0.951023   
2    0.208852        -0.682604               -1.575780         -1.357663   
3   -1.499940         1.349723               -0.664723         -0.544382   
4    0.493651        -0.005161                1.590516         -0.951023   
..        ...              ...                     ...               ...   
520  1.348047        -0.682604                2.053512          1.488821   
521  0.778450        -0.682604                0.171657         -0.679929   
522  1.348047         0.672281               -0.664723          1.421048   
523  0.778450        -1.360046               -0.664723          1.421048   
524 -1.499940        -0.682604                0.022303         -0.273288   

          Age  Daily Work Load Average  Body Mass Index  Children      Pets  \
0    2.105128                 2.662867         1.000046 -0.043001 -0.591608   
1   -1.340766                -0.862291        -0.418938 -0.946029 -0.591608   
2    0.068918                -1.047905         0.290554 -0.946029 -0.591608   
3    0.538813                 1.049692        -1.128430  0.860026 -0.591608   
4   -1.340766                -1.662075        -0.418938 -0.946029 -0.591608   
..        ...                      ...              ...       ...       ...   
520 -1.340766                -0.084581         0.054057 -0.043001  2.635345   
521  1.008707                -0.766032         2.655527 -0.043001 -0.591608   
522  0.225549                 0.868588         1.000046 -0.946029 -0.591608   
523  0.225549                -0.255001         1.000046 -0.946029 -0.591608   
524 -1.340766                 0.925235        -0.655435 -0.043001  1.021868   

     Reason1  Reason2  Reason3  Education  
0        0.0      0.0      0.0        0.0  
1        1.0      0.0      0.0        1.0  
2        0.0      0.0      0.0        0.0  
3        0.0      0.0      1.0        1.0  
4        1.0      0.0      0.0        1.0  
..       ...      ...      ...        ...  
520      0.0      0.0      0.0        0.0  
521      0.0      0.0      1.0        0.0  
522      0.0      0.0      0.0        0.0  
523      0.0      0.0      0.0        0.0  
524      0.0      0.0      0.0        0.0  

[525 rows x 13 columns]

The categorical feature is untouched, transform our test set.

In [186]:
X_test_scaled = ct.transform(X_test)
X_test_scaled = pd.DataFrame(X_test_scaled, columns = ['Month', 'Day_of_the_week',
       'Transportation Expense', 'Distance to Work', 'Age',
       'Daily Work Load Average', 'Body Mass Index', 'Children',
       'Pets','Reason1', 'Reason2', 'Reason3', 'Education'])
X_test_scaled

Month  Day_of_the_week  Transportation Expense  Distance to Work  \
0    0.493651        -1.360046               -0.664723          1.421048   
1    0.208852        -0.005161               -0.664723          1.421048   
2    1.063249         2.704608                0.545041          1.353274   
3   -0.645544        -0.005161               -1.575780         -1.357663   
4   -1.215141         0.672281               -0.664723          1.421048   
..        ...              ...                     ...               ...   
170 -0.930342        -0.005161               -1.575780         -1.154343   
171 -0.075946        -1.360046               -0.664723          1.421048   
172 -0.360745        -0.005161               -0.664723         -0.273288   
173  0.493651        -1.360046                2.053512          1.488821   
174  1.348047        -0.682604               -0.664723          1.421048   

          Age  Daily Work Load Average  Body Mass Index  Children      Pets  \
0    0.225549                -0.556353         1.000046 -0.946029 -0.591608   
1    0.225549                -0.814464         1.000046 -0.946029 -0.591608   
2   -0.087714                -0.172827        -0.891933  2.666081 -0.591608   
3    0.068918                -0.644775         0.290554 -0.946029 -0.591608   
4    0.225549                -0.192180         1.000046 -0.946029 -0.591608   
..        ...                      ...              ...       ...       ...   
170  2.105128                -0.692678         1.000046 -0.043001 -0.591608   
171  0.225549                -0.451526         1.000046 -0.946029 -0.591608   
172 -1.027503                -0.862291        -1.837922 -0.946029 -0.591608   
173 -1.340766                -1.662075         0.054057 -0.043001  2.635345   
174  0.225549                -0.084581         1.000046 -0.946029 -0.591608   

     Reason1  Reason2  Reason3  Education  
0        0.0      0.0      0.0        0.0  
1        0.0      0.0      0.0        0.0  
2        0.0      0.0      0.0        0.0  
3        1.0      0.0      0.0        0.0  
4        1.0      0.0      0.0        0.0  
..       ...      ...      ...        ...  
170      1.0      0.0      0.0        0.0  
171      0.0      0.0      0.0        0.0  
172      0.0      0.0      0.0        1.0  
173      1.0      0.0      0.0        0.0  
174      0.0      0.0      1.0        0.0  

[175 rows x 13 columns]

In [187]:
# Retrain
logreg = LogisticRegression()
logreg.fit(X_train_scaled, y_train)

LogisticRegression()

In [188]:
logreg.score(X_train_scaled, y_train)

0.72

In [189]:
# reinterpret
featurename = X_train_scaled.columns.values
sumtable = pd.DataFrame(columns = ['Feature Name'], data = featurename)
sumtable['Coef'] = np.transpose(logreg.coef_)
sumtable.index = sumtable.index+1
sumtable.loc[0] = ['Intercept', logreg.intercept_[0]]
sumtable.sort_index(inplace=True)
sumtable['Odds_ratio'] = np.exp(sumtable.Coef)
sumtable

Feature Name      Coef  Odds_ratio
0                 Intercept -0.690401    0.501375
1                     Month  0.114710    1.121548
2           Day_of_the_week -0.277519    0.757661
3    Transportation Expense  0.547303    1.728584
4          Distance to Work -0.017835    0.982323
5                       Age -0.188555    0.828155
6   Daily Work Load Average  0.085505    1.089267
7           Body Mass Index  0.103076    1.108575
8                  Children  0.331190    1.392625
9                      Pets -0.317793    0.727753
10                  Reason1  0.992718    2.698560
11                  Reason2  0.446541    1.562897
12                  Reason3  2.525047   12.491477
13                Education -0.019176    0.981006

We leave reason4 to be our baseline, reason4 are absent with no reason or light reason(dental, medical consult, etc).

So to re-interpret :
1. if employee absent with reason3 (injury and/or poisoning), that employee will 12.5 times likely to take excessive amount of absent compared to employee who absent with reason4, same with other reason with their respectiv odds ratio.
2. an employee with college-grad above will 0.98 times likely to take excessive absent
3. for numerical data, for each STANDARDIZE unit of that value, not unit of that values. so its not intepretable.

# More Feature Selections

refer back to intepretation, an odds_ratio that close to one is not really affecting the prediction of absent being excessive or not, because the change are very miniscule. we can eliminate them.

In [190]:
sumtable.sort_values(by='Odds_ratio')

Feature Name      Coef  Odds_ratio
0                 Intercept -0.690401    0.501375
9                      Pets -0.317793    0.727753
2           Day_of_the_week -0.277519    0.757661
5                       Age -0.188555    0.828155
13                Education -0.019176    0.981006
4          Distance to Work -0.017835    0.982323
6   Daily Work Load Average  0.085505    1.089267
7           Body Mass Index  0.103076    1.108575
1                     Month  0.114710    1.121548
8                  Children  0.331190    1.392625
11                  Reason2  0.446541    1.562897
3    Transportation Expense  0.547303    1.728584
10                  Reason1  0.992718    2.698560
12                  Reason3  2.525047   12.491477

In [200]:
df3 = df2.drop(['Distance to Work', 'Daily Work Load Average', 'Month'], axis=1)
df3

Reason1  Reason2  Reason3  Reason4  Day_of_the_week  \
0          0        0        0        1                1   
1          1        0        0        0                1   
2          0        0        0        1                2   
3          1        0        0        0                3   
4          0        0        0        1                3   
..       ...      ...      ...      ...              ...   
695        1        0        0        0                2   
696        1        0        0        0                2   
697        1        0        0        0                3   
698        0        0        0        1                3   
699        0        0        0        1                3   

     Transportation Expense  Age  Body Mass Index  Education  Children  Pets  \
0                       289   33               30          0         2     1   
1                       118   50               31          0         1     0   
2                       179   38               31          0         0     0   
3                       279   39               24          0         2     0   
4                       289   33               30          0         2     1   
..                      ...  ...              ...        ...       ...   ...   
695                     179   40               22          1         2     0   
696                     225   28               24          0         1     2   
697                     330   28               25          1         0     0   
698                     235   32               25          1         0     0   
699                     291   40               25          0         1     1   

     Target  
0         1  
1         0  
2         0  
3         1  
4         0  
..      ...  
695       1  
696       0  
697       1  
698       0  
699       0  

[700 rows x 12 columns]

In [203]:
X_train, X_test, y_train, y_test = train_test_split(df3.drop(['Target', 'Reason4'], axis=1), df3.Target, test_size = 0.25, stratify = df3.Target)

In [204]:
coltotrans = ['Day_of_the_week', 
              'Transportation Expense', 'Age', 
              'Body Mass Index', 'Children', 'Pets']
ct = make_column_transformer([StandardScaler(), coltotrans], remainder='passthrough')
X_train_scaled = ct.fit_transform(X_train)

In [205]:
X_train_scaled = pd.DataFrame(X_train_scaled, columns = ['Day_of_the_week',
       'Transportation Expense', 'Age', 'Body Mass Index', 'Children',
       'Pets', 'Reason1', 'Reason2', 'Reason3', 'Education'])
X_train_scaled

Day_of_the_week  Transportation Expense       Age  Body Mass Index  \
0          -0.650011               -1.544399  0.088406         0.298672   
1           1.357376               -0.629484  0.558809        -1.109737   
2          -1.319140               -0.989450 -0.381997        -0.405532   
3          -1.319140               -0.629484  0.245207         1.002877   
4          -1.319140               -0.629484  0.245207         1.002877   
..               ...                     ...       ...              ...   
520        -0.650011               -0.629484  0.245207         1.002877   
521         1.357376               -0.629484  0.558809        -1.109737   
522         0.019118               -0.629484  0.245207         1.002877   
523         0.688247                0.210437  0.088406         0.533407   
524         0.019118               -1.544399  2.126820         1.002877   

     Children      Pets  Reason1  Reason2  Reason3  Education  
0   -0.927783 -0.574465      0.0      0.0      0.0        0.0  
1    0.906816 -0.574465      0.0      0.0      0.0        1.0  
2    0.906816 -0.574465      1.0      0.0      0.0        0.0  
3   -0.927783 -0.574465      0.0      0.0      0.0        0.0  
4   -0.927783 -0.574465      0.0      0.0      0.0        0.0  
..        ...       ...      ...      ...      ...        ...  
520 -0.927783 -0.574465      0.0      0.0      0.0        0.0  
521  0.906816 -0.574465      0.0      0.0      1.0        1.0  
522 -0.927783 -0.574465      0.0      0.0      0.0        0.0  
523 -0.010483  0.277496      0.0      0.0      0.0        1.0  
524 -0.010483 -0.574465      1.0      0.0      0.0        0.0  

[525 rows x 10 columns]

In [206]:
X_test_scaled = ct.transform(X_test)
X_test_scaled = pd.DataFrame(X_test_scaled, columns = ['Day_of_the_week',
       'Transportation Expense', 'Age', 'Body Mass Index', 'Children',
       'Pets','Reason1', 'Reason2', 'Reason3', 'Education'])
X_test_scaled

Day_of_the_week  Transportation Expense       Age  Body Mass Index  \
0           0.688247                0.375422  0.715610        -0.875002   
1           0.019118                1.050359  0.558809        -0.405532   
2           0.019118                1.185346  1.029212        -0.405532   
3           0.019118                2.100261 -1.322804         0.063937   
4          -1.319140                1.050359  0.558809        -0.405532   
..               ...                     ...       ...              ...   
170        -0.650011                0.375422  0.715610        -0.875002   
171        -0.650011                1.020362 -0.538798         0.768142   
172        -1.319140                0.210437  0.088406         0.533407   
173         0.688247                1.050359  0.558809        -0.405532   
174         0.688247               -0.629484 -1.009202        -1.813942   

     Children      Pets  Reason1  Reason2  Reason3  Education  
0   -0.927783 -0.574465      1.0      0.0      0.0        0.0  
1   -0.010483  0.277496      0.0      0.0      0.0        0.0  
2    0.906816  0.277496      1.0      0.0      0.0        0.0  
3   -0.010483  2.833378      0.0      0.0      0.0        0.0  
4   -0.010483  0.277496      1.0      0.0      0.0        0.0  
..        ...       ...      ...      ...      ...        ...  
170 -0.927783 -0.574465      0.0      0.0      0.0        0.0  
171  0.906816  0.277496      0.0      0.0      0.0        0.0  
172 -0.010483  0.277496      1.0      0.0      0.0        1.0  
173 -0.010483  0.277496      0.0      0.0      0.0        0.0  
174 -0.927783 -0.574465      0.0      0.0      0.0        1.0  

[175 rows x 10 columns]

In [207]:
logreg = LogisticRegression()
logreg.fit(X_train_scaled, y_train)
logreg.score(X_train_scaled, y_train)

0.7047619047619048

The model perform slightly worse, from 0.72 to 0.704, however, a simpler model is always prefered, not only faster training (when we have huge amount of data it will impact the time), we also get simpler intepretation.

In [208]:
featurename = X_train_scaled.columns.values
sumtable = pd.DataFrame(columns = ['Feature Name'], data = featurename)
sumtable['Coef'] = np.transpose(logreg.coef_)
sumtable.index = sumtable.index+1
sumtable.loc[0] = ['Intercept', logreg.intercept_[0]]
sumtable.sort_index(inplace=True)
sumtable['Odds_ratio'] = np.exp(sumtable.Coef)
sumtable.sort_values(by='Odds_ratio')

Feature Name      Coef  Odds_ratio
0                Intercept -0.712597    0.490369
6                     Pets -0.314027    0.730499
1          Day_of_the_week -0.235868    0.789885
3                      Age -0.141938    0.867675
10               Education -0.108264    0.897391
4          Body Mass Index  0.082850    1.086378
8                  Reason2  0.218766    1.244540
5                 Children  0.403489    1.497038
2   Transportation Expense  0.455452    1.576886
7                  Reason1  1.180107    3.254722
9                  Reason3  1.777729    5.916402

The ratio changed but the intepretation is always the same.

with more analysis, more feature processes can be done. for example, day.

for increasing unit of day and month, the odds are decreasing (less than 1 odds_ratio) from the model perspective; absent will less likely to be excessive closer to weekend (5 which is friday, is lesser than 1, which is monday). with more consideration, we can decide to include or not, if it included, should we treat it as category rather than numerical feature?.

but for now lets use the model we trained and test it.

In [209]:
logreg.score(X_test_scaled, y_test)

0.7028571428571428

The model score on test data(Unseen data) is very close to the train score, this mean the training process yield a model that able to generalize.

by now, rather than getting the PREDICTED CLASS result, you might want to know the PREDICTED PROBABILITY.

In [221]:
pred_class = logreg.predict(X_test_scaled)
pred_proba = logreg.predict_proba(X_test_scaled)

In [217]:
pred_proba[:, 1] 
# These are the probability of absent being excessive, all probability below 0.5(by default) will be categorized as 0 
# while 0.5 above will be 1 which is excessive)

array([0.52700949, 0.39102542, 0.75062254, 0.3866362 , 0.7413035 ,
       0.26577393, 0.24058508, 0.5579948 , 0.74817644, 0.14445419,
       0.21293894, 0.24058508, 0.50936461, 0.22628194, 0.7413035 ,
       0.16400651, 0.16400651, 0.27947268, 0.54697559, 0.52874994,
       0.48592747, 0.46277708, 0.74175967, 0.22744303, 0.19085948,
       0.24058508, 0.22628194, 0.21642727, 0.28622817, 0.58256049,
       0.19085948, 0.21197475, 0.48235008, 0.18768453, 0.48592747,
       0.68107558, 0.24058508, 0.28622817, 0.24058508, 0.24882203,
       0.79357172, 0.31893819, 0.35415064, 0.21293894, 0.24095082,
       0.16400651, 0.28622817, 0.46874261, 0.34718703, 0.39102542,
       0.47642294, 0.25509894, 0.3547734 , 0.81367679, 0.42918131,
       0.61244626, 0.61244626, 0.60898969, 0.42971386, 0.19085948,
       0.14677945, 0.27947268, 0.48235008, 0.53516033, 0.38584315,
       0.78509294, 0.39600138, 0.18680755, 0.3866362 , 0.48539829,
       0.40806829, 0.66954517, 0.69858843, 0.25502984, 0.54152

# More Metrics

Accuracy alone is not enough metrics to evaluate the model, enter confusion matrix, we will take a look at model's each prediction compared to the actual label.

In [218]:
from sklearn.metrics import confusion_matrix, classification_report

In [228]:
confusion_matrix(y_test, pred_class)

array([[78, 17],
       [35, 45]], dtype=int64)

* The Top left is prediction 0, with actual label 0, A true Negatives
* The Bottom Right is prediction 1, with actual label 1, A true Positives
* The top right is prediction 1 with actual label 0, a false positive
* The bottom left is prediction 0 with actual label 1, a false negative

In [224]:
print(classification_report(y_test, pred_class))

              precision    recall  f1-score   support

           0       0.69      0.82      0.75        95
           1       0.73      0.56      0.63        80

    accuracy                           0.70       175
   macro avg       0.71      0.69      0.69       175
weighted avg       0.71      0.70      0.70       175



* Precision : How many positive prediction is correct prediction? we will have lower precision if we have more false positive. our model has 0.73 precision on positive class, that means, when it predict an employee taking an exesive amount of absent, it is correct 73% of the time
* Recall  : How many actual positive was indentified correctly? we will have lower recall if we have more false negative. our model has 0.56 recall on positive class, that means, the model correctly identified 56% of all employee on positive class(excessive amount of absent time)

In short, model with no false positive has precision of 1, and model with no false negative has recall of 1. If a business is impacted with false negative, in this case an employee that will take excessive amount of absent is predicted 0(will not take excessive amount of absent) the business might need a model with high recall, to do this we can adjust the prediction threshold, by default that threshold is 0.5.

In [236]:
adj_pred_class = np.where(pred_proba[:, 1] < 0.4, 0, 1)
#prediction more than or equal to 0.3 probability will be classified as positive class, will take excessive amount of time of absent
print(classification_report(y_test, adj_pred_class))

              precision    recall  f1-score   support

           0       0.78      0.75      0.76        95
           1       0.71      0.75      0.73        80

    accuracy                           0.75       175
   macro avg       0.75      0.75      0.75       175
weighted avg       0.75      0.75      0.75       175



Notice how the precision is now reduced, because the lower threshold, we now have more false positive.

# Save Model

Before saving the model, because to train the model we have to scale the data first, in other words, that scaling process is part of the model. for that we have to combine the scaling process with our logisticregression model, enter Pipeline.

In [241]:
#scaler is saved as ct and logisticregression is saved as logreg
from sklearn.pipeline import make_pipeline
pipe = make_pipeline(ct, logreg)

In [242]:
# test the model again, now with unscaled data because it will go through the included scaling process
pipe.score(X_test, y_test)

0.7028571428571428

In [245]:
pipe.predict_proba(X_test)

array([[0.47299051, 0.52700949],
       [0.60897458, 0.39102542],
       [0.24937746, 0.75062254],
       [0.6133638 , 0.3866362 ],
       [0.2586965 , 0.7413035 ],
       [0.73422607, 0.26577393],
       [0.75941492, 0.24058508],
       [0.4420052 , 0.5579948 ],
       [0.25182356, 0.74817644],
       [0.85554581, 0.14445419],
       [0.78706106, 0.21293894],
       [0.75941492, 0.24058508],
       [0.49063539, 0.50936461],
       [0.77371806, 0.22628194],
       [0.2586965 , 0.7413035 ],
       [0.83599349, 0.16400651],
       [0.83599349, 0.16400651],
       [0.72052732, 0.27947268],
       [0.45302441, 0.54697559],
       [0.47125006, 0.52874994],
       [0.51407253, 0.48592747],
       [0.53722292, 0.46277708],
       [0.25824033, 0.74175967],
       [0.77255697, 0.22744303],
       [0.80914052, 0.19085948],
       [0.75941492, 0.24058508],
       [0.77371806, 0.22628194],
       [0.78357273, 0.21642727],
       [0.71377183, 0.28622817],
       [0.41743951, 0.58256049],
       [0.

In [247]:
X_test.shape

(175, 10)

In [248]:
X_test

Reason1  Reason2  Reason3  Day_of_the_week  Transportation Expense  Age  \
345        1        0        0                3                     246   41   
556        0        0        0                2                     291   40   
256        1        0        0                2                     300   43   
399        0        0        0                2                     361   28   
87         1        0        0                0                     291   40   
..       ...      ...      ...              ...                     ...  ...   
496        0        0        0                1                     246   41   
0          0        0        0                1                     289   33   
439        1        0        0                0                     235   37   
71         0        0        0                3                     291   40   
460        0        0        0                3                     179   30   

     Body Mass Index  Education  Children  Pets  
345               23          0         0     0  
556               25          0         1     1  
256               25          0         2     1  
399               27          0         1     4  
87                25          0         1     1  
..               ...        ...       ...   ...  
496               23          0         0     0  
0                 30          0         2     1  
439               29          1         1     1  
71                25          0         1     1  
460               19          1         0     0  

[175 rows x 10 columns]

In [258]:
X_train.columns

Index(['Reason1', 'Reason2', 'Reason3', 'Day_of_the_week',
       'Transportation Expense', 'Age', 'Body Mass Index', 'Education',
       'Children', 'Pets'],
      dtype='object')

In [254]:
X_train_scaled

Day_of_the_week  Transportation Expense       Age  Body Mass Index  \
0          -0.650011               -1.544399  0.088406         0.298672   
1           1.357376               -0.629484  0.558809        -1.109737   
2          -1.319140               -0.989450 -0.381997        -0.405532   
3          -1.319140               -0.629484  0.245207         1.002877   
4          -1.319140               -0.629484  0.245207         1.002877   
..               ...                     ...       ...              ...   
520        -0.650011               -0.629484  0.245207         1.002877   
521         1.357376               -0.629484  0.558809        -1.109737   
522         0.019118               -0.629484  0.245207         1.002877   
523         0.688247                0.210437  0.088406         0.533407   
524         0.019118               -1.544399  2.126820         1.002877   

     Children      Pets  Reason1  Reason2  Reason3  Education  
0   -0.927783 -0.574465      0.0      0.0      0.0        0.0  
1    0.906816 -0.574465      0.0      0.0      0.0        1.0  
2    0.906816 -0.574465      1.0      0.0      0.0        0.0  
3   -0.927783 -0.574465      0.0      0.0      0.0        0.0  
4   -0.927783 -0.574465      0.0      0.0      0.0        0.0  
..        ...       ...      ...      ...      ...        ...  
520 -0.927783 -0.574465      0.0      0.0      0.0        0.0  
521  0.906816 -0.574465      0.0      0.0      1.0        1.0  
522 -0.927783 -0.574465      0.0      0.0      0.0        0.0  
523 -0.010483  0.277496      0.0      0.0      0.0        1.0  
524 -0.010483 -0.574465      1.0      0.0      0.0        0.0  

[525 rows x 10 columns]

the model is now complete, save it

In [243]:
import pickle

In [244]:
with open('model', 'wb') as file:
    pickle.dump(pipe, file)

In [257]:
logreg.n_iter_

array([25])